drafts/client-interface.md

```diff
@@ -3,11 +3,150 @@
 This document describes the interface of a yet-to-be-built reference client
 library in pseudo code.
 
+The hydra client works similar to a browser. You point it to a URL. The URL becomes the current resource. 
+Then you interact with the controls of the resource.
 
-## Retrieve a resource
+The examples show how a cli application could use the hydra client for interaction with an API.
+
+## Find API entrypoint for website
+
+Hydra has a special mechanism to discover API entrypoints.
+
+```
+hydra.findEntrypoint(website)
+print hydra.representation 
+// the API entrypoint is now the current representation
+```
+
+## Retrieve a resource from URI
+
+### Using uniform interface
+
+Assuming that apiurl is known as API entrypoint and that it is a 
+[FoodEstablishment](http://schema.org/FoodEstablishment). The menu information is not embedded, but there is only a link to the menu, 
+i.e. the FoodEstablishment representation has a control to retrieve the menu. 
+
+This assumes that the client is aware of the uniform interface of the underlying protocol. 
+In the example below the client knows that it uses the uniform interface of HTTP.
+```
+hydra.setLocation(apiurl);
+// apiurl is now the current representation
+
+// we want to retrieve or download the menu of the food establishment
+menu = hydra.controls["http://schema.org/hasMenu"]
+if(menu && menu.methods["GET"])
+    hydra.execute("GET", menu.href)
+    print hydra.representation
+    hydra.back()
```

Comment:
> Let me argue that we should make hydraClient a browser of sorts. If you work with an API, you visit several representations until you finally reach your goal.

For me, a browser is an app. I envisioned the Hydra client to be something lower level. There's no better or worse, they are just different levels of abstraction.

> If you have moved to a different representation, you are usually not interested in the previous representations anymore. Then why should the user of the hydraClient hold references to all those representations?

I'd argue that depends entirely on the app, i.e., the goal you are trying to achieve

> Can you please illustrate your point that there may be multiple branches or error handling which are difficult to handle if there is only one current representation and the hydraClient holds it?

Imagine I pass `hydra` to a method which in turn passes it to another method and then I continue. What's the state of `hydra`? I'd need to introspect it to be sure. How should it behave if an operation fails (network error e.g.)? Should it stay at the previous state or move internally to an error state?